src/QirRuntime/lib/Tracer/tracer.cpp

```diff
@@ -0,0 +1,305 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <cassert>
+#include <set>
+#include <sstream>
+
+#include "tracer.hpp"
+
+using namespace std;
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    thread_local std::shared_ptr<CTracer> tracer = nullptr;
+    std::shared_ptr<CTracer> CreateTracer(int preferredLayerDuration)
+    {
+        tracer = std::make_shared<CTracer>(preferredLayerDuration);
+        return tracer;
+    }
+    std::shared_ptr<CTracer> CreateTracer(int preferredLayerDuration, const std::unordered_map<OpId, std::string>& opNames)
+    {
+        tracer = std::make_shared<CTracer>(preferredLayerDuration, opNames);
+        return tracer;
+    }
+
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer's ISimulator implementation
+    //------------------------------------------------------------------------------------------------------------------
+    Qubit CTracer::AllocateQubit()
+    {
+        size_t qubit = qubits.size();
+        qubits.emplace_back(QubitState{});
+        return reinterpret_cast<Qubit>(qubit);
+    }

```

Comment:
May be in the future we'll consider using an integer type of fixed size (like `int32_t`) as a `Qubit` (and its `.ll` counterpart). 
E.g. in `CoreTypes.hpp`:
`using Qubit = int32_t;`
It will be assumed that `Qubit` type is a 32-bit integer descriptor/handle of the actual allocated qubit (and documented like that).

Such a change will let us avoid all the casts between the pointer type and integer type (including in `.ll` bridge). Everywhere it will be passed natively as `Qubit` or `i32`.

Why of _fixed size_? Because some day we will need to pass this `Qubit` descriptor between the machines/architectures/clusters/datacenters.

Reply:
That's feedback on the QIR spec. The qubit handle should probably be i64 to allow implementers to stuff a pointer into it, if they'd wish